File: examples/dotnet/TransactionService.cs
using System.Text;
using System.Runtime.InteropServices;

using WebAssembly;
using WebAssembly.Runtime;
using Newtonsoft.Json;
using Models;

public class TransactionService {
    private WebAssembly.Instance<dynamic> serviceInstance;

    public TransactionService(Stream wasmModule) {
        Instance<dynamic> instance = Compile.FromBinary<dynamic>(wasmModule)(new ImportDictionary());
        this.serviceInstance = instance;
    }

    public TransactionService(System.String wasmModulePath) : this(File.OpenRead(wasmModulePath)) {}

    public InformationResponse information() {
        Func<int, int> function = x => this.serviceInstance.Exports.information(x);
        return this.callWasmFunction<Object, InformationResponse>(
            new object(),
            function
        )!;
    }
    
    public Manifest convert_manifest(
        ConvertManifestRequest request
    ) {
        Func<int, int> function = x => this.serviceInstance.Exports.convert_manifest(x);
        return this.callWasmFunction<Object, Manifest>(
            request,
            function
        )!;
    }
    
    public CompileTransactionIntentResponse compile_transaction_intent(
        TransactionIntent request
    ) {
        Func<int, int> function = x => this.serviceInstance.Exports.compile_transaction_intent(x);
        return this.callWasmFunction<Object, CompileTransactionIntentResponse>(
            request,
            function
        )!;
    }
    
    public TransactionIntent decompile_transaction_intent(
        DecompileTransactionIntentRequest request
    ) {
        Func<int, int> function = x => this.serviceInstance.Exports.decompile_transaction_intent(x);
        return this.callWasmFunction<Object, TransactionIntent>(
            request,
            function
        )!;
    }

    public CompileSignedTransactionIntentResponse compile_signed_transaction_intent(
        SignedTransactionIntent request
    ) {
        Func<int, int> function = x => this.serviceInstance.Exports.compile_signed_transaction_intent(x);
        return this.callWasmFunction<Object, CompileSignedTransactionIntentResponse>(
            request,
            function
        )!;
    }
    
    public SignedTransactionIntent decompile_signed_transaction_intent(
        DecompileSignedTransactionIntentRequest request
    ) {
        Func<int, int> function = x => this.serviceInstance.Exports.decompile_signed_transaction_intent(x);
        return this.callWasmFunction<Object, SignedTransactionIntent>(
            request,
            function
        )!;
    }

    public CompileNotarizedTransactionIntentResponse compile_notarized_transaction_intent(
        NotarizedTransactionIntent request
    ) {
        Func<int, int> function = x => this.serviceInstance.Exports.compile_notarized_transaction_intent(x);
        return this.callWasmFunction<Object, CompileNotarizedTransactionIntentResponse>(
            request,
            function
        )!;
    }

    public NotarizedTransactionIntent decompile_notarized_transaction_intent(
        DecompileNotarizedTransactionIntentRequest request
    ) {
        Func<int, int> function = x => this.serviceInstance.Exports.decompile_notarized_transaction_intent(x);
        return this.callWasmFunction<Object, NotarizedTransactionIntent>(
            request,
            function
        )!;
    }
    
    public DecompileUnknownTransactionIntentResponse decompile_unknown_transaction_intent(
        DecompileUnknownTransactionIntentRequest request
    ) {
        Func<int, int> function = x => this.serviceInstance.Exports.decompile_unknown_transaction_intent(x);
        return this.callWasmFunction<Object, DecompileUnknownTransactionIntentResponse>(
            request,
            function
        )!;
    }

    private Resp callWasmFunction<Req, Resp>(
        Req request,
        Func<int, int> func
    ) {
        // Allocate memory and perform the function call
        IntPtr requestMemoryLocation = this.writeRequestToMemory(request!);
        IntPtr responseMemoryLocation = (IntPtr)func((int)requestMemoryLocation);
        Resp response = this.readResponseFromMemory<Resp>(responseMemoryLocation)!;

        // Free up the allocated memory
        this.freeMemory(requestMemoryLocation);
        this.freeMemory(responseMemoryLocation);

        return response;
    }

    private IntPtr writeRequestToMemory(Object request) {
        // Serialize the passed object and add a null terminator to the end of it.
        System.String serializedRequest = JsonConvert.SerializeObject(request);
        serializedRequest += '\0';

        // Encoding it and allocating enough memory for it
        byte[] encodedRequest = Encoding.UTF8.GetBytes(serializedRequest);
        IntPtr allocatedMemoryLocation = this.allocateMemory((UInt32)encodedRequest.Length);

        // Writing the string to memory
        Marshal.Copy(
            encodedRequest, 
            0, 
            this.serviceInstance.Exports.memory.Start + allocatedMemoryLocation.ToInt32(), 
            encodedRequest.Length
        );

        // Return a pointer to the object in memory
        return allocatedMemoryLocation;
    }

    private T? readResponseFromMemory<T>(IntPtr pointer) {
        // Read the string from memory until the null terminator
        IntPtr memoryLocation = this.serviceInstance.Exports.memory.Start + pointer.ToInt32();
        System.String? responseString = Marshal.PtrToStringAnsi(memoryLocation);
        if (responseString == null) {
            throw new Exception("Response string is null");
        } else {
            return JsonConvert.DeserializeObject<T>(responseString);
        }
    }

    private IntPtr allocateMemory(UInt32 capacity) {
        return (IntPtr)this.serviceInstance.Exports.__transaction_lib_alloc((int)capacity);
    }
    
    private void freeMemory(IntPtr pointer) {
        this.serviceInstance.Exports.__transaction_lib_free((int)pointer);
    }
}

File: examples/typescript/src/transaction-service.ts
import {
	CompileTransactionIntentRequest,
	CompileTransactionIntentResponse,
	ConvertManifestRequest,
	ConvertManifestResponse,
	DecompileTransactionIntentRequest,
	DecompileTransactionIntentResponse,
	InformationRequest,
	InformationResponse,
	Request,
	Response,
	TransactionServiceInterface,
	CompileSignedTransactionIntentRequest,
	CompileSignedTransactionIntentResponse,
	DecompileSignedTransactionIntentRequest,
	DecompileSignedTransactionIntentResponse,
	CompileNotarizedTransactionIntentRequest,
	CompileNotarizedTransactionIntentResponse,
	DecompileNotarizedTransactionIntentRequest,
	DecompileNotarizedTransactionIntentResponse,
	DecompileUnknownTransactionIntentRequest,
	DecompileUnknownTransactionIntentResponse,
	EncodeAddressRequest,
	EncodeAddressResponse,
	DecodeAddressRequest,
	DecodeAddressResponse,
	SBOREncodeRequest,
	SBOREncodeResponse,
	SBORDecodeRequest,
	SBORDecodeResponse,
} from "./interfaces";
import { Error } from "./error";

export default class TransactionAPI {
	private wasmModule: WebAssembly.Instance;
	private internal_service: TransactionServiceInterface;

	// =============
	// Constructors
	// =============

	constructor(wasmModule: WebAssembly.Instance) {
		this.wasmModule = wasmModule;
		this.internal_service = wasmModule.exports as unknown as TransactionServiceInterface;
	}

	static async fromWasmModuleBuffer(buffer: Uint8Array): Promise<TransactionAPI> {
		let wasmImports: WebAssembly.Imports = {
			env: {
				memory: new WebAssembly.Memory({
					initial: 1,
				}),
			},
		};
		let wasmModule: WebAssembly.WebAssemblyInstantiatedSource = await WebAssembly.instantiate(
			buffer,
			wasmImports
		);
		return new TransactionAPI(wasmModule.instance);
	}

	// =================
	// Exported Methods
	// =================

	information(): InformationResponse | Error {
		return this.callWasmFunction({} as InformationRequest, this.internal_service.information) as
			| InformationResponse
			| Error;
	}

	convertManifest(request: ConvertManifestRequest): ConvertManifestResponse | Error {
		return this.callWasmFunction(request, this.internal_service.convert_manifest) as
			| ConvertManifestResponse
			| Error;
	}

	compileTransactionIntent(
		request: CompileTransactionIntentRequest
	): CompileTransactionIntentResponse | Error {
		return this.callWasmFunction(request, this.internal_service.compile_transaction_intent) as
			| CompileTransactionIntentResponse
			| Error;
	}

	decompileTransactionIntent(
		request: DecompileTransactionIntentRequest
	): DecompileTransactionIntentResponse | Error {
		return this.callWasmFunction(request, this.internal_service.decompile_transaction_intent) as
			| DecompileTransactionIntentResponse
			| Error;
	}

	compileSignedTransactionIntent(
		request: CompileSignedTransactionIntentRequest
	): CompileSignedTransactionIntentResponse | Error {
		return this.callWasmFunction(
			request,
			this.internal_service.compile_signed_transaction_intent
		) as CompileSignedTransactionIntentResponse | Error;
	}

	decompileSignedTransactionIntent(
		request: DecompileSignedTransactionIntentRequest
	): DecompileSignedTransactionIntentResponse | Error {
		return this.callWasmFunction(
			request,
			this.internal_service.decompile_signed_transaction_intent
		) as DecompileSignedTransactionIntentResponse | Error;
	}

	compileNotarizedTransactionIntent(
		request: CompileNotarizedTransactionIntentRequest
	): CompileNotarizedTransactionIntentResponse | Error {
		return this.callWasmFunction(
			request,
			this.internal_service.compile_notarized_transaction_intent
		) as CompileNotarizedTransactionIntentResponse | Error;
	}

	decompileNotarizedTransactionIntent(
		request: DecompileNotarizedTransactionIntentRequest
	): DecompileNotarizedTransactionIntentResponse | Error {
		return this.callWasmFunction(
			request,
			this.internal_service.decompile_notarized_transaction_intent
		) as DecompileNotarizedTransactionIntentResponse | Error;
	}

	decompileUnknownTransactionIntent(
		request: DecompileUnknownTransactionIntentRequest
	): DecompileUnknownTransactionIntentResponse | Error {
		return this.callWasmFunction(
			request,
			this.internal_service.decompile_unknown_transaction_intent
		) as DecompileUnknownTransactionIntentResponse | Error;
	}

	encodeAddress(request: EncodeAddressRequest): EncodeAddressResponse | Error {
		return this.callWasmFunction(request, this.internal_service.encode_address) as
			| EncodeAddressResponse
			| Error;
	}

	decodeAddress(request: DecodeAddressRequest): DecodeAddressResponse | Error {
		return this.callWasmFunction(request, this.internal_service.decode_address) as
			| DecodeAddressResponse
			| Error;
	}

	sborEncode(request: SBOREncodeRequest): SBOREncodeResponse | Error {
		console.log(request);
		return this.callWasmFunction(request, this.internal_service.sbor_encode) as
			| SBOREncodeResponse
			| Error;
	}

	sborDecode(request: SBORDecodeRequest): SBORDecodeResponse | Error {
		return this.callWasmFunction(request, this.internal_service.sbor_decode) as
			| SBORDecodeResponse
			| Error;
	}

	private callWasmFunction(
		request: Request,
		wasmFunction: (requestStringPointer: number) => number
	): Response | Error {
		// Serialize the request as JSON and write it to WASM's memory
		let requestStringPointer: number = this.writeString(JSON.stringify(request));

		// Call the method on the WASM module
		let responsePointer: number = wasmFunction(requestStringPointer);

		// Read and parse the returned response
		let returnedString: string = this.readString(responsePointer);
		let parsedResponse: Response | Error = JSON.parse(returnedString);

		// Free up the memory needed in this operation
		this.internal_service.__transaction_lib_free(requestStringPointer);
		this.internal_service.__transaction_lib_free(responsePointer);

		return parsedResponse;
	}

	// ===============
	// Helper Methods
	// ===============

	private readString(pointer: number): string {
		// @ts-ignore
		let memoryBuffer: Uint8Array = this.wasmModule.exports.memory.buffer;

		const view: Uint8Array = new Uint8Array(memoryBuffer, pointer);
		const length: number = view.findIndex((byte) => byte === 0);
		const decoder: TextDecoder = new TextDecoder();

		return decoder.decode(new Uint8Array(memoryBuffer, pointer, length));
	}

	private writeString(string: string): number {
		const pointer: number = this.allocateMemory(string);

		// @ts-ignore
		let memoryBuffer: Uint8Array = this.wasmModule.exports.memory.buffer;

		const view: Uint8Array = new Uint8Array(memoryBuffer, pointer);
		const encoder: TextEncoder = new TextEncoder();
		view.set(new Uint8Array([...encoder.encode(string), 0])); // Adding 0 at the end to be a c-style string

		return pointer;
	}

	private allocateMemory(string: string): number {
		// Take the string and convert it into a byte array to determine its length
		let byteArray: Uint8Array = new TextEncoder().encode(string);
		let pointer: number = this.internal_service.__transaction_lib_alloc(byteArray.length + 1);
		return pointer;
	}
}
